### PR TITLE
Add new ICMP and UDP packet analyzers

### PR DESCRIPTION
**Pull request by timwoj**
This continues the work merged in #1516. This PR adds a new framework to the packet analyzers for IP-Based analyzers, as well as UDP and ICMP analyzers based on that framework. Most of the changes to the analyzers themselves are just code moving from one place to another. I have a few comments myself for some sticky spots, but this work passes all of tests and the benchmarks are comparable to what is currently in master.

TCP will come in a later PR, as it’s probably going to be complicated enough to warrant a PR just for it.

### CODE REVIEW

File: src/packet_analysis/protocol/ip/IPBasedAnalyzer.h [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+	                            const u_char* data, bool& flip_roles) const
+		{
+		flip_roles = false;
+		return true;
+		}
+
+	/**
+	 * Returns the port mask for an analyzer. Used by IsLikelyServerPort().
+	 */
+	virtual uint32_t GetServerPortMask() const = 0;
```
Comment (timwoj) [inline src/packet_analysis/protocol/ip/IPBasedAnalyzer.h:88]: This method and the method below just return a constant from the child classes and feel like they could just be replaced with a couple of member variables. A pure virtual method here forces them to be overridden though where a variable could be missed/forgotten about.

File: scripts/base/packet-protocols/tcp/main.zeek [hunk 2]
```diff
@@ -0,0 +1,5 @@
+module PacketAnalyzer::TCP;
+
+event zeek_init() &priority=20
```
Comment (timwoj) [inline scripts/base/packet-protocols/tcp/main.zeek:3]: The `zeek_init` events in these files probably should be commented out if they're not doing anything. No use adding yet another event to be called that's just a no-op.

File: src/packet_analysis/protocol/ip/IPBasedAnalyzer.cc [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+		conn->SetRecordPackets(true);
+		conn->SetRecordContents(true);
+
+		const u_char* data = pkt->ip_hdr->Payload();
+
+		run_state::current_timestamp = run_state::processing_start_time;
+		run_state::current_pkt = pkt;
+
+		// TODO: Does this actually mean anything?
+		if ( conn->Skipping() )
```
Comment (timwoj) [inline src/packet_analysis/protocol/ip/IPBasedAnalyzer.cc:85]: This TODO here exists because according to the following comment in Conn.h, skipping isn't handled there anymore:

```
// FIXME: Now this is in Analyzer and should eventually be removed here.
```

Cleaning that up felt a bit out of scope for this PR though, so I left it as a TODO.

File: src/packet_analysis/protocol/ip/IPBasedAnalyzer.h [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+	 * session manager. This should be implemented by all child classes.
+	 *
+	 * @param conn The connection currently being processed.
+	 * @param t The timestamp for the current packet.
+	 * @param is_orig Flag denoting whether this packet is from the originator of
+	 * the connection.
+	 * @param remaining The remaining about of data in the packet.
+	 * @param pkt The packet being processed.
+	 */
+	virtual void ContinueProcessing(Connection* conn, double t, bool is_orig, int remaining,
```
[the remark this answers is not in the document]
Comment (rsmmr) [reply src/packet_analysis/protocol/ip/IPBasedAnalyzer.h:115]: More generally, the control flow feels hard to follow here: AnalyzerPacket -> ProcessConnection -> ContinueProcessing -> DeliverPacket --- this jumps back and forth between the derived and base class. It would be more easy to follow if one side stayed in charge of the flow and just called the other for help. One idea with the base class as the driver:

1. Implement `AnalyzePacket()` in the base class, have it call a virtual method `BuildConnTuple`, then directly continue with what's in `ProcessConnection`.
2. Once it comes to `ContinueProcessing`, rename that a virtual method `DeliverPacket` that does the dispatching to the transport layer analyzer (and everything else). 
3. Then also pass the current `session_analyzer` around with the method calls, maintaining the member variable seems error prone. 

Alternatively, flip it around: implement `AnalyzerPacket()` inside the derived class and call the base call for help as needed; might make maintaining the `session_analyzer` more natural. not sure.

File: src/packet_analysis/protocol/ip/IPBasedAnalyzer.h [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+	zeek::Connection* NewConn(const ConnTuple* id, const detail::ConnKey& key,
+	                          const Packet* pkt);
+};
+
+/**
+ * This class represents the interface between the packet analysis framework and
+ * the session analysis framework. One of these should be implemented for each
+ * packet analyzer that intends to forward into the session analysis.
+ */
+class IPBasedTransportAnalyzer : public zeek::analyzer::TransportLayerAnalyzer {
```
Comment (rsmmr) [inline src/packet_analysis/protocol/ip/IPBasedAnalyzer.h:147]: I'm wondering if this should move into `analyzer/`. That would fit semantically, and it would put into the analyzer namespace as well, which would make it quite a bit more readable I think. Right now, we have `packet_analysis::IP::IPBasedAnalyzer`, which is a packet analyzer; and then `packet_analysis::IP::IPBasedTransportAnalyzer`, which is a connection analyzer -- that's feels to track.

The next question then would be if should split up UDP/ICMP/TCP similarly. I'm inclined to say yes, but not fully sure.

**Review by rsmmr — COMMENTED** [review]
I'm not fully through yet, but starting with a set of comments mostly about the new IP analyzer API.

File: scripts/base/packet-protocols/ip/main.zeek [hunk 6]
```diff
@@ -1,8 +1,18 @@
...
 event zeek_init() &priority=20
 	{
 	PacketAnalyzer::register_packet_analyzer(PacketAnalyzer::ANALYZER_IP, 4, PacketAnalyzer::ANALYZER_IPTUNNEL);
 	PacketAnalyzer::register_packet_analyzer(PacketAnalyzer::ANALYZER_IP, 41, PacketAnalyzer::ANALYZER_IPTUNNEL);
 	PacketAnalyzer::register_packet_analyzer(PacketAnalyzer::ANALYZER_IP, 47, PacketAnalyzer::ANALYZER_GRE);
+
+	PacketAnalyzer::register_packet_analyzer(PacketAnalyzer::ANALYZER_IP, IPPROTO_TCP, PacketAnalyzer::ANALYZER_TCP_PKT);
+	PacketAnalyzer::register_packet_analyzer(PacketAnalyzer::ANALYZER_IP, IPPROTO_UDP, PacketAnalyzer::ANALYZER_UDP_PKT);
+	PacketAnalyzer::register_packet_analyzer(PacketAnalyzer::ANALYZER_IP, IPPROTO_ICMP, PacketAnalyzer::ANALYZER_ICMP_PKT);
+	PacketAnalyzer::register_packet_analyzer(PacketAnalyzer::ANALYZER_IP, IPPROTO_ICMP6, PacketAnalyzer::ANALYZER_ICMP_PKT);
```
[the remark this answers is not in the document]
Comment (timwoj) [reply scripts/base/packet-protocols/ip/main.zeek]: This gets fixed in a later commit in the branch.

File: src/packet_analysis/protocol/icmp/ICMP.h [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+	bool AnalyzePacket(size_t len, const uint8_t* data, Packet* packet) override;
+
+	static zeek::packet_analysis::AnalyzerPtr Instantiate()
+		{
+		return std::make_shared<ICMPAnalyzer>();
+		}
+
+private:
+
```
[the remark this answers is not in the document]
Comment (timwoj) [reply src/packet_analysis/protocol/icmp/ICMP.h:119]: Same as above. The `private:` block gets fleshed out in all of those classes later. I could remove them from that first commit just for cleanliness, I suppose.

[the remark this answers is not in the document]
Comment (rsmmr) [reply src/packet_analysis/protocol/icmp/ICMP.h:119]: No need, I just went through the commits in order. :-)

File: src/packet_analysis/protocol/ip/IPBasedAnalyzer.h (hunk 5 above)
Comment (timwoj) [reply src/packet_analysis/protocol/ip/IPBasedAnalyzer.h:140]: I think the problem here is really naming and not location. The "transport analyzers" here aren't really analyzers in the pure sense. They're really just a glue layer between packet analysis and the child analyzers that used to be in the old analyzers. They're not intended to be called by anyone other than the packet analyzers they're associated with, so I'm sort of leery to move them.

File: src/packet_analysis/protocol/ip/IPBasedAnalyzer.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (timwoj) [reply src/packet_analysis/protocol/ip/IPBasedAnalyzer.h]: The comments here should all be resolved now. I found a few other things that we didn't need to be storing during processing as well, especially in the ICMP analyzer.

File: src/packet_analysis/protocol/ip/IPBasedAnalyzer.h [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+	 * terminated the previous one, i.e., the new data is arriving at
+	 * what's the analyzer for the previous instance. This is used only
+	 * for TCP.
+	 */
+	virtual bool IsReuse(double t, const u_char* pkt) { return false; }
+
+	/**
+	 * TODO: comment
+	 */
+	virtual void CreateTransportAnalyzer(Connection* conn, IPBasedTransportAnalyzer*& root,
```
[the remark this answers is not in the document]
Comment (timwoj) [reply src/packet_analysis/protocol/ip/IPBasedAnalyzer.h]: I moved the code out of `analyzer::Manager` and reworked things a little bit in cb72668d19fb2abfaa6f12353b6cbf0442c092fd. I'm debating whether moving the port registration maps over into the packet analyzers (and splitting them up) makes sense. I think the registration methods can stay but the actual maps can move. `analyzer::Manager` would have to cache pointers to the UDP/TCP analyzers to do the registration/unregistration but that wouldn't be a huge change.

[the remark this answers is not in the document]
Comment (rsmmr) [reply src/packet_analysis/protocol/ip/IPBasedAnalyzer.h]: yeah, sounds reasonable.

**Comment by rsmmr** [comment]
let me know when you're done with the next set of updates and I'll take another look (thinking about any renaming in particular).

File: src/packet_analysis/protocol/ip/IPBasedAnalyzer.h (hunk 5 above)
Comment (rsmmr) [reply src/packet_analysis/protocol/ip/IPBasedAnalyzer.h:138]: Ok, I can see this being primarily an issue of naming.

File: src/packet_analysis/protocol/ip/IPBasedAnalyzer.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (timwoj) [reply src/packet_analysis/protocol/ip/IPBasedAnalyzer.h]: Scratch that, I'm going to hold off on this until I do the TCP analyzer so I can move all of it at once. I'm also not sure how I want to store the analyzers there in the manager yet, since I don't really want to add more lookups against `TransportProto` when I'm trying to remove as much of that as I can. That said, all of those registrations/lookups are based on the `PortProto()` value from a `PortVal`, which is a fixed enum anyways.

File: src/packet_analysis/protocol/icmp/ICMP.h [hunk 9]
```diff
@@ -17,7 +17,7 @@ using RecordValPtr = IntrusivePtr<RecordVal>;
 
 namespace packet_analysis::ICMP {
 
-class ICMPTransportAnalyzer;
+class ICMPAnalyzerAdapter;
```
Comment (rsmmr) [inline src/packet_analysis/protocol/icmp/ICMP.h:20]: Given that we're in the ICMP namespace now, can skip the ICMP prefix I'd say (similar elsewhere)

File: src/packet_analysis/protocol/ip/IP.h [hunk 10]
```diff
@@ -32,4 +32,28 @@ class IPAnalyzer : public Analyzer {
...
+ *        to hold an IP header.
+ * @return 0 If the inner IP packet appeared valid, else -1 if \a caplen
+ *         is greater than the supposed IP packet's payload length field, -2
+ *         if the version of the inner header does not match proto or
+ *         1 if \a caplen is less than the supposed packet's payload length.
+ *         In the -1 case, \a inner may still be non-null if \a caplen was
+ *         long enough to be an IP header, and \a inner is always non-null
+ *         for other return values.
+ */
+int ParsePacket(int caplen, const u_char* const pkt, int proto,
```
Comment (rsmmr) [inline src/packet_analysis/protocol/ip/IP.h:57]: Can this be `static`?

File: src/packet_analysis/protocol/ip/AnalyzerAdapter.h [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+namespace zeek::packet_analysis::IP {
+
+class IPBasedAnalyzer;
+
+/**
+ * This class represents the interface between the packet analysis framework and
+ * the session analysis framework. One of these should be implemented for each
+ * packet analyzer that intends to forward into the session analysis.
+ */
+class AnalyzerAdapter : public analyzer::Analyzer {
```
Comment (rsmmr) [inline src/packet_analysis/protocol/ip/AnalyzerAdapter.h:16]: sorry but one more naming idea for this: how about `SessionAdaptor`? that would make the tie in with the session framework apparent.

File: src/packet_analysis/protocol/ip/IPBasedAnalyzer.h [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+#include "zeek/packet_analysis/protocol/ip/AnalyzerAdapter.h"
+#include "zeek/analyzer/Analyzer.h"
+#include "zeek/analyzer/Manager.h"
+
+namespace zeek::analyzer::pia { class PIA; }
+
+namespace zeek::packet_analysis::IP {
+
+/**
+ * A base class for any packet analyzer based on IP. This is used by default by
```
Comment (rsmmr) [inline src/packet_analysis/protocol/ip/IPBasedAnalyzer.h:16]:
```suggestion
 * A base class for reuse by packet analyzers based on IP. This is used by default by
```

To make clear it's optional to use this (assuming it is?)

File: src/packet_analysis/protocol/ip/IPBasedAnalyzer.h [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+	 *
+	 * @param port The port number to check, in host order.
+	 */
+	bool IsLikelyServerPort(uint32_t port) const;
+
+	// TODO: temporary, until all of the plugins are implemented
+	bool new_plugin = false;
+
+	TransportProto transport;
+	uint32_t server_port_mask;
```
Comment (rsmmr) [inline src/packet_analysis/protocol/ip/IPBasedAnalyzer.h:115]: More public member variables, better to avoid (see comment elsewhere).

File: src/packet_analysis/Analyzer.h [hunk 14]
```diff
@@ -94,6 +94,11 @@ class Analyzer {
 	 */
 	void RegisterProtocol(uint32_t identifier, AnalyzerPtr child);
 
+	/**
+	 * Flag for whether to report unknown protocols in ForwardPacket.
+	 */
+	bool report_unknown_protocols = true;
```
[the remark this answers is not in the document]
Comment (rsmmr) [reply src/packet_analysis/Analyzer.h:100]: Ah, looking at usage, can also pass into constructor I think.

File: src/packet_analysis/protocol/ip/IPBasedAnalyzer.h [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+	 * is used to hook into the session analyzer framework. This function can also be used
+	 * to do any extra initialization of connection timers, etc.
+	 */
+	virtual AnalyzerAdapter* MakeAnalyzerAdapter(Connection* conn) { return nullptr; }
+
+	/**
+	 * Returns a PIA appropriate for this IP-based analyzer.
+	 */
+	virtual analyzer::pia::PIA* MakePIA(Connection* conn) { return nullptr; }
+
```
Comment (rsmmr) [inline src/packet_analysis/protocol/ip/IPBasedAnalyzer.h:89]: Add a comment to these two saying whether derived class must implement these and return instances. `MakeAnalyzerAdapter` probably needs to be, but we should be ok without a PIA?

Comment (rsmmr) [reply src/packet_analysis/protocol/ip/IPBasedAnalyzer.h:89]: Oh, and if one of the two *must* be implemented make it pure. Maybe the absence of that means even the `MakeAnalyzerAdapter` is optional?

File: src/session/Manager.cc [hunk 16]
```diff
@@ -390,6 +159,17 @@ Connection* Manager::FindConnection(Val* v)
 	return conn;
 	}
 
+Connection* Manager::FindConnection(const zeek::detail::ConnKey& conn_key)
+	{
+	detail::Key key(&conn_key, sizeof(conn_key), false);
+
+	auto it = session_map.find(key);
+	if ( it != session_map.end() )
+		return static_cast<Connection*>(it->second);
```
Comment (rsmmr) [inline src/session/Manager.cc:168]: Hmm, can we use `dynamic_cast` here to double-check it's actually a `Connection`? Not sure about performance, but might be just fine. If it wasn't, we could at least still do that in debug compiles?

**Review by rsmmr — APPROVED** [review]
LGTM, just some more small stuff.

One additional testing piece: I don't know how well we have recording of packet/session content covered in our test-suite. Because there are some changes coming with this work to how/when that is done, can you check if we got the main cases covered and, where not, add tests ensuring we don't change semantics of what gets written to disk?

File: src/packet_analysis/protocol/ip/AnalyzerAdapter.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (timwoj) [reply src/packet_analysis/protocol/ip/AnalyzerAdapter.h:16]: I was going off the ol' Design Patterns, which calls it `Adapter`.

File: src/packet_analysis/protocol/ip/IP.h (hunk 10 above)
Comment (timwoj) [reply src/packet_analysis/protocol/ip/IP.h:57]: As in static within the IPAnalyzer class? I'm indifferent either way.

File: src/packet_analysis/protocol/ip/IPBasedAnalyzer.h (hunk 15 above)
Comment (timwoj) [reply src/packet_analysis/protocol/ip/IPBasedAnalyzer.h:89]: `MakeAnalyzerAdapter` can be made pure virtual, but not until the TCP work is done. I'll add a comment to note that it needs to be fixed before that work is done.

File: src/packet_analysis/protocol/ip/IP.h (hunk 10 above)
Comment (rsmmr) [reply src/packet_analysis/protocol/ip/IP.h:57]: Nevermind, I misread the diff, thought it was a method.

File: src/packet_analysis/protocol/ip/AnalyzerAdapter.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (rsmmr) [reply src/packet_analysis/protocol/ip/AnalyzerAdapter.h:16]: okey-dokey. :-)

**Review by rsmmr — APPROVED** [review]

File: src/packet_analysis/protocol/ip/AnalyzerAdapter.h (hunk 11 above)
Comment (timwoj) [reply src/packet_analysis/protocol/ip/AnalyzerAdapter.h:16]: Seems reasonable to me. I'll squash that back into the `Adapter` commit before I merge this.

File: src/session/Manager.cc (hunk 16 above)
Comment (timwoj) [reply src/session/Manager.cc]: This was the performance difference I DM'd you about. Changing this to a `dynamic_cast` is actually fairly costly.

**Comment by timwoj** [comment]
This was for #1114 (adding a reference here so it shows up in the issue).
